In [3]:
print('Hello,\vworld')

Hello,world


# Escape sequences

'\b' is ASCII Backspace. Just look at this!

In [14]:
x = 'Hello,'
y = 'world!'
print('Default is:', x + y)
print('Modified is:', x +'\b' + ' ' + y + '\b')

Default is: Hello,world!
Modified is: Hello, world!


'\r\ is ASCII carriage return. It overwrites N symbols at the end of the string where N is length of the string after this '\r' sign.

In [18]:
print('Hello, world!')
print('Hello, \rworld')

Hello, world!
world


# String methods

capitalize() returns a string with first letter capitalized. 

In [41]:
str = 'Hello123'

In [22]:
print(str.capitalize())

Hdjhggjdgrjsghkgykfjhgu57^ugtuyjjkjhfjkhgykfhjkghgghkgk


center() return a string which is padded with the specified character. It takes 2 arguments: width and char.

In [24]:
print(str.center(100, '_'))

______________________hdjhggjdgrJSGhkgykfJHgU57^uGTuyJJkjhFjkHGykfhjkGhgGhkGk_______________________


casefold() converts strings to casefolded strings. Look, ß becomes ss but lower() doesn't touch it as it's lowercase.

In [26]:
print('der Fluß'.casefold())

der fluss


count() returns the number of occurences of a substring in the given string.

In [28]:
print(str.count('g'))

6


endswith() checks if a string ends with the specified suffix.

In [30]:
print(str.endswith('k'))

True


The expandtabs() method returns a copy of string with all tab characters '\t' replaced with whitespace characters until the next multiple of tabsize parameter.

In [31]:
print('abcd\t12345678\txy'.expandtabs())

abcd    12345678        xy


Default tabsize is 8. Length of 'abcd' is 4. So we get 8 - 4 = 4 spaces after it. Length of '12345678' is 8. So we get 8 * 2 - 8 = 8 spaces after it. Length of 'xy' is only 2. And we get 6 spaces after it. Got it?

encode() can convert unicoded strings into any encodings supported by Python. It takes 2 arguments: encoding and errors. By default, Python uses utf-8 encoding. 

In [33]:
print('pythön'.encode('ascii', 'ignore'))

b'pythn'


The find() method returns the index of first occurrence of the substring (if found). If not found, it returns -1.

The find() method takes maximum of three parameters:
    1) sub - It's the substring to be searched in the str string.
    2) start and end (optional) - substring is searched within str[start:end]

In [36]:
print(str.find('g', 2, 14))
print(str.find('МО', 2, 14))

4
-1


The rfind() method returns the highest index of the substring (if found). If not found, it returns -1.

In [78]:
a = 'cat dog cat'
a.rfind('cat')

8

index() does the same as find() but it raises and exception if the substring is not found.

In [37]:
print(str.index('g', 2, 14))
print(str.index('МО', 2, 14))

4


ValueError: substring not found

rindex() does the same as rfind() but it raises and exception if the substring is not found.

In [79]:
a = 'cat dog cat'
a.rindex('cat')

8

The isalnum() method returns True if all characters in the string are alphanumeric (either alphabets or numbers).

In [42]:
print(str.isalnum())
print('!@&'.isalnum())

True
False


The isalpha() method returns True if all characters in the string are alphabets.

In [44]:
print(str.isalpha())
print('asd'.isalpha())

False
True


The isdecimal() method returns True if all characters in a string are decimal characters. Superscripts and subscripts (usually written using unicode) aren't considered decimal characters.

In [45]:
print(str.isdecimal())
print('123456'.isdecimal())

False
True


The isdigit() method returns True if all characters in a string are digits. Superscripts and subscripts (usually written using unicode) are considered digit characters.

In [47]:
print(str.isdigit())
print('123456'.isdigit())

False
True


The isidentifier() method returns True if the string is a valid identifier in Python.

In [48]:
print('x1'.isidentifier())
print('1x'.isidentifier())

True
False


The islower() method returns True if all alphabets in a string are lowercase alphabets.

In [ ]:
print('Abc'.islower())
print('abc'.islower())

The isnumeric() method returns True if all characters in a string are numeric characters.

In Python, decimal characters (like: 0, 1, 2..), digits (like: subscript, superscript), and characters having Unicode numeric value property (like: fraction, roman numerals, currency numerators) are all considered numeric characters.

In [52]:
print('123'.isnumeric())
print('½'.isnumeric())

True
True


The isprintable() methods returns True if all characters in the string are printable or the string is empty.

In [54]:
print('abc xyz'.isprintable())
print('abc\nxyz'.isprintable())

True
False


The isspace() method returns True if there are only whitespace characters in the string.

In [55]:
'\n\t'.isspace()

True

The istitle() returns True if the string is a titlecased string.

In [56]:
'Python Is Awesome'.istitle()

True

The string isupper() method returns whether or not all characters in a string are uppercased or not.

In [57]:
'PYTHON'.isupper()

True

The join() is a string method which returns a string concatenated with the elements of an iterable.

In [59]:
'_'.join(['it', 'works', 'like', 'this'])

'it_works_like_this'

The string ljust() method returns a left-justified string of a given minimum width.

The ljust() method takes two parameters:

    1) width - width of the given string. If width is less than or equal to the length of the string, original string is returned.
    
    2) fillchar (Optional) - character to fill the remaining space of the width

In [60]:
str.ljust(50, '_')

'Hello123__________________________________________'

rjust() works the same way but return a right-justified string.

In [61]:
str.rjust(50, '_')

'__________________________________________Hello123'

The string lower() method converts all uppercase characters in a string into lowercase characters and returns it.

In [62]:
str.lower()

'hello123'

The string upper() method converts all lowercase characters in a string into uppercase characters and returns it.

In [63]:
str.upper()

'HELLO123'

The string swapcase() method converts all uppercase characters to lowercase and all lowercase characters to uppercase characters of the given string, and returns it.

In [64]:
str.swapcase()

'hELLO123'

The lstrip() method returns a copy of the string with leading characters removed (based on the string argument passed).

In [69]:
a = '    dgdfgdfgd  '
b = '______gdfgfdg___'
print(a.lstrip(' '))
print(b.lstrip('_'))

dgdfgdfgd  
gdfgfdg___


The lstrip() method returns a copy of the string with trailing characters removed (based on the string argument passed).

In [70]:
a = '    dgdfgdfgd  '
b = '______gdfgfdg___'
print(a.rstrip(' '))
print(b.rstrip('_'))

    dgdfgdfgd
______gdfgfdg


strip() removes Both Leading and Trailing Characters

In [71]:
a = '    dgdfgdfgd  '
b = '______gdfgfdg___'
print(a.strip(' '))
print(b.strip('_'))

dgdfgdfgd
gdfgfdg


The partition() method splits the string at the first occurrence of the argument string and returns a tuple containing the part the before separator, argument string and the part after the separator.

In [73]:
a = '    dgdfgdfgd  '
a.partition('f')

('    dgd', 'f', 'gdfgd  ')

The rpartition() splits the string at the last occurrence of the argument string and returns a tuple containing the part the before separator, argument string and the part after the separator.

In [75]:
a = '    dgdfgdfgd  '
a.rpartition('f')

('    dgdfgd', 'f', 'gd  ')

maketrans() function is used to construct the transition table i.e specify the list of characters that need to be replaced in the whole string or the characters that need to be deleted from the string.

Parameters :

    1) str1 : Specifies the list of characters that need to be replaced.
    2) str2 : Specifies the list of characters with which the characters need to be replaced.
    3) str3 : Specifies the list of characters that needs to be deleted.

Returns : Returns the translation table which specifies the conversions that can be used by translate()

To translate the characters in the string translate() is used to make the translations. 

This function uses the translation mapping specified using the maketrans().

Parameters :

    1) table : Translate mapping specified to perform translations.
    2) delstr : The delete string can be specified as optional argument is not mentioned in table.

Returns : Returns the argument string after performing the translations using the translation table.

In [121]:
a = ''

# specify to translate chars 
str1 = "wy"
  
# specify to replace with 
str2 = "gf"
  
# delete chars 
str3 = "u"
  
# target string  
trg = "weeksyourweeks"
  
# using maketrans() to  
# construct translate 
# table 
table = a.maketrans(str1, str2, str3) 
  
# Printing original string  
print ("The string before translating is : ", end ="") 
print (trg) 
  
# using translate() to make translations. 
print ("The string after translating is : ", end ="") 
print (trg.translate(table)) 

NameError: name 'maketrans' is not defined

The replace() method returns a copy of the string where all occurrences of a substring is replaced with another substring.

The replace() method can take maximum of 3 parameters:
    1) old - old substring you want to replace
    2) new - new substring which would replace the old substring
    3) count (optional) - the number of times you want to replace the old substring with the new substring

In [83]:
a = 'cat cat cat'
a.replace('cat', 'dog', 2)

'dog dog cat'

The split() method breaks up a string at the specified separator and returns a list of strings.

The split() method takes maximum of 2 parameters:
    
    1) separator (optional)- The is a delimiter. The rsplit() method splits string starting from the left at the specified separator.
    
If the separator is not specified, any whitespace (space, newline etc.) string is a separator.

    2) maxsplit (optional) - The maxsplit defines the maximum number of splits.
    
The default value of maxsplit is -1, meaning, no limit on the number of splits.

In [84]:
a.split(' ')

['cat', 'cat', 'cat']

The rsplit() method splits string from the right at the specified separator and returns a list of strings.

The split() method takes maximum of 2 parameters:
       
    1) separator (optional)- The is a delimiter. The rsplit() method splits string starting from the right at the specified separator.
    
If the separator is not specified, any whitespace (space, newline etc.) string is a separator.

    2) maxsplit (optional) - The maxsplit defines the maximum number of splits.

The default value of maxsplit is -1, meaning, no limit on the number of splits.

We can see difference between split() and rsplit() only when we specify the 'maxsplit' argument.

In [85]:
a.split(' ')

['cat', 'cat', 'cat']

The splitlines() method splits the string at line breaks and returns a list of lines in the string.

The splitlines() takes maximum of 1 parameter.

    1) keepends (optional) - If keepends is provided and True, line breaks are also included in items of the list.

By default, the line breaks are not included.

In [89]:
a = """cat
dog
cat
goose
pig
"""
print(a.splitlines())
print(a.splitlines(True))

['cat', 'dog', 'cat', 'goose', 'pig']
['cat\n', 'dog\n', 'cat\n', 'goose\n', 'pig\n']


The startswith() method returns True if a string starts with the specified prefix(string).

The startswith() method takes maximum of three parameters:

    1) prefix - String or tuple of strings to be checked
    
    2) start (optional) - Beginning position where prefix is to be checked within the string.
    
    3) end (optional) - Ending position where prefix is to be checked within the string.

In [100]:
a = 'abc def ghi'
print(a.startswith('bc def', 1, 5))
print(a.startswith('bc def', 1))

False
True


The title() method returns a string with first letter of each WORD capitalized; a title cased string.

NB! capitalize() returns a string with the first letter of the STRING capitalized.

In [103]:
a.title()

'Abc Def Ghi'

The zfill() method returns a copy of the string with '0' characters padded to the left.

If a string starts with the sign prefix('+', '-'), the '0' digits are filled after the first sign prefix character.

In [110]:
print(a.zfill(20))
print('+300'.zfill(10))

000000000abc def ghi
+000000300


Python string format_map() function is very similar to format() function. 

The only difference is that the mappings are used directly and not copied to a dictionary. 

This is useful when the mapping is a dict subclass.

In [114]:
s = 'My name is {name} and I am a {job_title}'

my_dict = {'name': 'Alexander', 'job_title': 'Student'}

s.format(**my_dict)

'My name is Alexander and I am a Student'

In [117]:
class MyDict(dict):
    def __missing__(self, key):
        return '___'


s = 'My name is {name} and I am a {job_title}'
my_dict = MyDict(name='Alexander')

print(s.format_map(my_dict))
print(s.format(**my_dict))  # gives an error

My name is Alexander and I am a ___


KeyError: 'job_title'